[2기-P] 최인창 SpringBoot Board JPA Mission 제출

ICCHOI opened the pull request:
# 게시판 만들기

## 요구사항

### Spring Data JPA를 설정한다

- [x]  datasource : h2 or MySQL

### **엔티티를 구성한다**

- [x]  회원(User)
    - id (PK) (auto increment)
    - name
    - age
    - hobby
    - **created_at (mapped superclass)**
    - **created_by (mapped superclass)**
- [x]  게시글(Post)
    - id (PK) (auto increment)
    - title
    - content
    - **created_at (mapped superclass)**
    - **created_by (mapped superclass)**
- [x]  회원과 게시글에 대한 연관관계를 설정한다.
    - 회원과 게시글은 1:N 관계이다.
    - mappedBy
    - cascade
    - orphan
- [x]  게시글 Repository를 구현한다. (PostRepository)

### **API를 구현한다.**

- [x]  게시글 조회
    - 페이징 조회 (GET "/posts")
    - 단건 조회 (GET "/posts/{id}")
- [x]  게시글 작성 (POST "/posts")
- [x]  게시글 수정 (POST "/posts/{id}")

### **REST-DOCS를 이용해서 문서화한다.**

- [x]  API 문서화

## TO DO :

- [ ] Pageable을 PageRequest로 변경합니다.

- [ ] BaseEntity의 creted_by를 Security의 인증과 연관시켜 구현합니다.

- [ ] 에러나 예외 상황에 대한 테스트 케이스 검증을 추가 합니다.

### 고민 중 :

Converter는 패키지의 어느 부분에 위치하는게 맞을까? 
PostStubService 중 Pageable을 사용한 테스트가 실패하는데 어떤 부분을 놓쳤는지
entity에 대한 검증을 어떤 방식으로 구현을 해야할까?
 - `@Valid`? 아니면 내부 메소드를 통한 검증?

ICCHOI committed:
Config:환경설정을 구성하라

ICCHOI committed:
init:프로젝트를 구성하라

ICCHOI committed:
test:테스트 코드를 작성하라

ICCHOI committed:
docs:API 문서화를 진행하라

ICCHOI committed:
feat:전역적으로 사용되는 에러를 구현하라

ICCHOI committed:
feat:user 도메인을 구현하라

ICCHOI committed:
feat:공통 도메인을 구현하라

ICCHOI committed:
feat:post api를 구현하라

ICCHOI committed:
feat:post domain을 구현하라

ICCHOI committed:
feat:post dto를 구현하라

ICCHOI committed:
feat:post dao를 구현하라

ICCHOI committed:
feat:post service를 구현하라

ICCHOI committed:
Config:애플리케이션 설정을 구현하라

zxcv9203 commented on src/main/java/org/programmers/springboardjpa/domain/model/BaseEntity.java:
```diff
@@ -0,0 +1,23 @@
+
+@Getter
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
```
해당 어노테이션은 어떤 역할을 하나요? 🤔

ICCHOI replied:
`@EntityListeners(AuditingEntityListener.class)`를 사용하여 BaseEntity에 audting 기능을 추가할 수 있습니다!
이 설정을 통해 `@CreatedDate`가 붙어 있는 곳에 해당 Entity가 생성 될 때 시간이 자동으로 저장 됩니다.

zxcv9203 commented on src/main/java/org/programmers/springboardjpa/domain/post/api/PostApi.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.List;
+
+@Valid
+@RequestMapping("/api/v1")
```
`/posts` 도 중복되는 url인 것 같은데 따로 매핑해주는 이유는 무엇인가요? 🤔

ICCHOI replied:
post의 각각의 엔드 포인트를 메소드 위에 매핑 해주면 좀 더 명시적으로 해당 api에 대해 표현 할 수 있다고 생각해서
따로 매핑하였습니다. :)

zxcv9203 commented on src/main/java/org/programmers/springboardjpa/domain/post/service/PostDefaultService.java:
```diff
@@ -0,0 +1,60 @@
+
+@Service
+@Transactional(readOnly = true)
+public class PostDefaultService implements PostService {
```
PostService의 경우 다른 구현체를 사용하지 않는데 인터페이스를 만드신 이유는 무엇인가요? 🤔

ICCHOI replied:
용철님 말씀대로 하나의 인터페이스에 하나의 구현체가 존재하기 때문에 postService를 굳이 나누지 않을 수도 있을 것 같습니다.
저는 결합도를 낮춘다는 측면과 후에 서비스가 비대해 질 경우를 생각해서 이런 구조를 고려 했었는데 서비스가 커질 경우는 기능별로 분리하는게 더 좋을 수도 있을 것 같네요!

SeokRae replied:
데브코스 시작부터 계속 나오는 고민인 것 같습니다.
이전에는 두었으나 현재는 인터페이스를 두지 않아도 되는 이유로 

```text
이전에는 JDK Dynamic Proxy를 통해 인터페이스를 기반으로 프록시 객체를 생성하였던 것이
최근 Spring 3.2 부터는 CGLIB을 사용하기 때문에 인터페이스를 굳이 두지 않더라도 AOP 구현이 가능하다. 
그래서 인터페이스를 두지 않아도 된다. 
```
라는 의견도 있긴 합니다만

하지만 인터페이스의 역할은 그것에만 있는 것이 아니라
Controller에서는 Service의 구현 객체를 알 필요 없고, 구현 방법이나 전략, 내부적으로 사용하는 또 다른 오브젝트에 대해서도 신경쓰지 않아도 사용할 수 있도록 인터페이스를 통해 낮은 결합도로 최소한으로 연결 시키는 것이 좋지 않을까 하는 생각입니다.

단순하게 네트워크 계층을 예시로 네트워크 통신의 경우 여러 기능을 계층으로 분리하여 추상화 단계를 제공하고 있습니다.
이론적으로 각 계층은 인접한 위아래 계층하고만 통신하게 되고, 네트워크를 몇 개의 계층으로 분리함으로써 계층 간 인터페이스가 변경되지 않는 한 하나의 계층 안에 있는 소프트웨어를 다른 계층에 미치는 영향 없이 수정하거나 대체할 수 있습니다.

이러한 방식으로 구현하게 되면 보다 안전하고 유지보수성이 높은 코드를 작성할 수 있지 않을까 생각됩니다.

ynoolee replied:
@SeokRae  궁금한 것이 있는데 현재처럼 작은 크기의 프로젝트이면서 구현체가 하나만 존재하는 경우에도 이렇게 인터페이스를 분리하는 것이 무조건 더 좋은것인지 궁금합니다. 

스프링 프레임워크에서는 객체 생성을 컨테이너에서 해주고 있기 때문에 내부적으로 사용하는 또 다른 오브젝트에 대해서는 클라이언트 측에서는 신경 쓰지 않을 수 있다는 생각이 들었고 
클래스를 통한 캡슐화 만으로도 사용하는 측에서는 구현에 대해서는 신경 쓰지 않을 수 있지 않을까? 라는 생각이 들었습니다

물론 이후의 확장성을 고려한다면 인터페이스를 통해 낮은 결합도로 , 객체간의 최소한의 연결을 하는 것이 필요하다는 생각이 들었습니다. 하지만  확장성만을 위해서 지금처럼 다른 구현체가 없는 상황에도 항상 인터페이스를 작성해주어야 하는 것이 맞을까? 라는 생각을 했습니다.  현재의 경우에는 하나의 PostService 에 기능이 계속해서 추가될 것 같은데, 그렇다면 인터페이스와 구현체 모두에 계속해서 추가하는 상황만 일어나지 않나 라는 생각이 듭니다. 만약 이렇게 인터페이스를 반드시 두겠다면  인터페이스나 구현체 에는 조금 더 의미 있는 이름이 있어야 하지 않을까? 라는 생각이 들었습니다

zxcv9203 commented on src/main/java/org/programmers/springboardjpa/global/config/MvcConfig.java:
```diff
@@ -0,0 +1,15 @@
+public class MvcConfig implements WebMvcConfigurer {
+
+    @Override
+    public void addCorsMappings(CorsRegistry registry) {
```
CORS 설정을 한 이유는 무엇인가요? 🤔

ICCHOI replied:
반복된 학습을 하며 여러 Config를 설정해보고 싶었습니다.

zxcv9203 commented on src/main/java/org/programmers/springboardjpa/global/error/ErrorResponse.java:
```diff
@@ -0,0 +1,85 @@
+    private String code;
+
+
+    private ErrorResponse(final ErrorCode code, final List<FieldError> errors) {
```
생성자에서 받는 파라미터를 final로 지정하신 이유가 궁금합니다 🤔

ICCHOI replied:
모든 변경 가능성이 없는 파라미터에 final을 붙임으로써 코드의 불변성을 보장해줄 수 있습니다.
엔티티의 필드를 설계할 때 기본적으로 final로 설계하고 변경에 소요가 있는 필드만 final을 빼는 것 처럼,
파라미터에도 이러한 원칙을 적용하면 디버깅이나 유지보수, 가독성에 이점이 있다고 생각 합니다

zxcv9203 commented on src/main/java/org/programmers/springboardjpa/global/error/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,55 @@
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler({MethodArgumentNotValidException.class, ConstraintViolationException.class})
+    protected ResponseEntity<ErrorResponse> handleMethodArgumentNotValidException(MethodArgumentNotValidException e) {
```
ApiResponse로 따로 응답 객체를 만들지 않고 ResponseEntity를 사용하신 이유가 궁금합니다 🤔

ICCHOI replied:
ApiResponse와 같은 Custom Response를 사용하는 것도 고려 해봤었는데 
아직은 서비스의 크기가 작아서 따로 만들 필요성 보다는 ResponseEntity로 충분하다고 생각했습니다.

zxcv9203 commented on src/main/java/org/programmers/springboardjpa/global/error/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,55 @@
+
+    @ExceptionHandler({MethodArgumentNotValidException.class, ConstraintViolationException.class})
+    protected ResponseEntity<ErrorResponse> handleMethodArgumentNotValidException(MethodArgumentNotValidException e) {
+        log.warn("handleMethodArgumentNotValidException", e);
```
log 레벨을 warn 레벨로 진행하신 이유가 무엇인가요? 🤔

ICCHOI replied:
비즈니스 에러들은 해결 가능한 에러이기 때문에 warn을 통해 관리 해야한다고 생각했습니다.

zxcv9203 commented on src/test/java/org/programmers/springboardjpa/post/service/PostDefaultServiceTest.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    @Test
+    @Transactional
```
mock 객체를 이용해서 테스트 하는 것 같은데 실제 DB에 데이터가 안들어가는 것 아닌가요? 🤔 @Transactional을 붙이신 이유가 궁금합니다 🤔

ICCHOI replied:
맞습니다! 필요 없는 어노테이션이네요. 제거 하였습니다 :)

zxcv9203 commented on pom.xml:
```diff
@@ -0,0 +1,106 @@
+
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-devtools</artifactId>
```
spring-boot-devtools는 어떤 기능이 필요하여 의존성을 추가하신건가요?

ICCHOI replied:
spring-boot-devtools는 개발 편의를 도와주는 모듈인데요, 코드에 수정 사항이 있을 때 서버가 자동으로 시작되는 Automatic Restart기능을 사용하려고 추가하였습니다. :)

SeokRae replied:
코드의 수정은 자바 코드 수정도 포함인가요?

ICCHOI replied:
자바 코드 수정 후 컴파일만 하면 서버를 다시 띄우지 않아도 자동으로 진행 해주는 것으로 알고 있습니다.
voucher 프로젝트의 의존성을 참고 하며 진행 했었는데, 이 프로젝트는 사실 api로 진행하고 따로 view가 없어
굳이 없어도 될 것 같네요! 제거 하는 게 좋을 것 같습니다 :)

SeokRae replied:
바로 적용되는 것은 classpath 하위에 있는 정적 리소스에 대해 변경이 발생하면 그를 적용해주는 것 같고
자바 코드의 경우, 설정 시 프로젝트 재빌드(Automatic Restart)가 되는 것 같네요.

참고: [Overview of Spring Boot Dev Tools](https://www.baeldung.com/spring-boot-devtools)

zxcv9203 commented on src/main/java/org/programmers/springboardjpa/domain/model/BaseEntity.java:
```diff
@@ -0,0 +1,23 @@
+@Getter
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+public abstract class BaseEntity {
```
BaseEntity가 추상클래스인 이유는 무엇인가요? 🤔

ICCHOI replied:
BaseEntity 같은 경우는 실제로 생성 될 일도 없고 자식 클래스에게 매핑 정보만 제공하는 역할을 하기 떄문에
추상 클래스로 작성해야한다고 생각했습니다.

SeokRae replied:
추상 클래스의 경우 직접 객체를 생성하지 못하기 때문에 
직접 생성을 제한하고 상속을 받아 사용하게끔 추상 클래스로 구현했다. 
`정리하면 단순 매핑 정보를 상속할 목적으로만 사용하기 때문에 추상메서드로 구현했다.` 라는 이유도 가능하겠군요.

zxcv9203 reviewed (approved):
JPA 게시판 미션 고생하셨습니다 🎉🎉🎉

SeokRae commented on src/main/java/org/programmers/springboardjpa/domain/post/api/PostApi.java:
```diff
@@ -0,0 +1,43 @@
+import javax.validation.constraints.Positive;
+import java.util.List;
+
+@Valid
```
클래스에 @Valid를 붙이면 어떻게 동작하는 걸까요?

ICCHOI replied:
Valid가 아니라 Validated를 붙혀 해당 클래스 내에서 hibernate-validator의 기능을 동작하도록 만들려고 했는데
 다른 어노테이션을 붙혔던 것 같습니다. 수정 하겠습니다 ;)

SeokRae commented on src/main/java/org/programmers/springboardjpa/domain/post/domain/Post.java:
```diff
@@ -0,0 +1,66 @@
+
+        Post post = (Post) o;
+
+        return new EqualsBuilder().append(id, post.id).isEquals();
```
오 new EqualsBuilder() 라는 클래스는 처음 보는데 Objects.equals(id, post.id)와 동일한 역할을 하는 건가요?

ICCHOI replied:
Security 강의를 통해서 commons lang과 guava에 대해 알게 되어 해당 라이브러리에 대해 공부하면서 라이브러리에서 지원하는 equals and hashcode 구현체를 사용해보았습니다.

SeokRae commented on src/main/java/org/programmers/springboardjpa/domain/post/dto/PostResponse.java:
```diff
@@ -0,0 +1,12 @@
+
+import org.programmers.springboardjpa.domain.user.dto.UserDto;
+
+public class PostResponse {
```
response dto를 PostResponse에 따로 두신 이유가 있을까요?
추가적으로 응답 별 dto가 생길 수 있기 때문에 request와 통일성있게 작성하신건가요?

ICCHOI replied:
넵 맞습니다! 요청에 따라 특정 api가 계속 생길 수 있기 때문에 분리하였습니다 :)

SeokRae commented on src/main/java/org/programmers/springboardjpa/domain/post/exception/NoExistIdException.java:
```diff
@@ -0,0 +1,10 @@
+import org.programmers.springboardjpa.global.error.exception.BusinessException;
+import org.programmers.springboardjpa.global.error.exception.ErrorCode;
+
+public class NoExistIdException extends BusinessException {
```
수정해야 하는 내용은 아니지만 Exception 네이밍은 기존의 표준 예외를 참고해보시는 것도 좋습니다.

SeokRae commented on src/main/java/org/programmers/springboardjpa/domain/post/service/PostConverter.java:
```diff
@@ -0,0 +1,46 @@
+import org.programmers.springboardjpa.domain.user.dto.UserDto.UserResponse;
+import org.springframework.stereotype.Component;
+
+@Component
```
새로 객체를 생성하는 용도의 클래스는 빈으로 등록해야 하는 걸까요?
인창님 의견 한번 들어보고 싶습니다.

ICCHOI replied:
Bean으로 관리 되어야하는 객체와 그렇지 않은 객체에 대해 고민을 많이 해 봤었는데 
결국 Covnerter 클래스는 불변 객체로 외부에서 값을 받아 변환하는 역할만 담당하고 있으며,
하나의 인스턴스만 생성하여 공유하는 것이 더 효율적이기 때문에 빈으로 등록 하였습니다.

SeokRae commented on src/main/java/org/programmers/springboardjpa/domain/post/service/PostDefaultService.java:
```diff
@@ -0,0 +1,60 @@
+    @Override
+    @Transactional
+    public PostResponseDto savePost(PostCreateRequestDto createRequest) {
+        Post post = postConverter.convertPost(createRequest);
```
스트림 로직으로 수정해보시는 것도 좋을 것 같습니다.

SeokRae commented on src/main/java/org/programmers/springboardjpa/domain/post/service/PostDefaultService.java:
```diff
@@ -0,0 +1,60 @@
+    @Override
+    @Transactional
+    public PostResponseDto updatePost(Long id, PostUpdateRequestDto updateRequest) {
+        var post = postRepository.findById(id).orElseThrow(NoExistIdException::new);
```
예외가 발생할 가능성이 있는 메서드가 findById 밖에 없는 거면, 이또한 스트림으로 변경 가능할 것 같습니다.
한번 도전해보시죵!

ynoolee replied:
위에서는 Post 타입을 사용하셨는데 여기서는 갑자기 var 을 사용하신 이유가 있으실까요?

ICCHOI replied:
따로 이유가 있었던 건 아니고, 자동생성으로 만들면 var로 생성되게 해서 이렇게 된 것 같은데
통일성을 위해 Post로 수정 하겠습니다!

SeokRae commented on src/main/java/org/programmers/springboardjpa/global/error/exception/ErrorCode.java:
```diff
@@ -0,0 +1,36 @@
+
+public enum ErrorCode {
+
+    POST_NOT_FOUND(404,"P001","존재하지 않는 게시글입니다."),
```
404 또는 500이 HttpStatus.NOT_FOUND를 의도한 것이라면 상수 그대로를 넣어도 좋을 듯하네요. !

ICCHOI replied:
좋은 방법이네요!  int status를 HttpStatus enum으로 대체 가능할 것 같습니다.

SeokRae commented on src/test/java/org/programmers/springboardjpa/post/service/PostDefaultServiceTest.java:
```diff
@@ -0,0 +1,127 @@
+
+@Transactional(readOnly = true)
+@ExtendWith(MockitoExtension.class)
+class PostDefaultServiceTest {
```
mockito를 이용한 단위 테스트를 진행하시는 것처럼 보이는데 @Transactional을 붙이신 이유가 있을까요?

ICCHOI replied:
굳이 필요없는 어노테이션이네요! 제거 하였습니다 ;)

SeokRae reviewed (approved):
이제 점점 피드백 할 부분이 줄어드네요. 🥺

커멘트 생각나는 것들 작성해보았습니다.

추가로 예외 처리의 경우, 왠지 같은 응답 코드의 경우에는 묶어서 처리할 수 있지 않을까 싶네요.
한번 고민해보시죠!

> 궁금한점

- Converter는 패키지의 어느 부분에 위치하는게 맞을까?
  - Converter를 사용하는 레이어에 가깝게 두는 것이 좋지 않을까 싶네요.
- PostStubService 중 Pageable을 사용한 테스트가 실패하는데 어떤 부분을 놓쳤는지
  - `Pageable.ofSize(1).withPage(1)`의 의미가 뭘까요?
  - `List<PostResponse.PostResponseDto> postList = postService.getPostList(PageRequest.of(0, 1));` 로 테스트 하면 정상 통과를 하는데 한번  해보시죵
- entity에 대한 검증을 어떤 방식으로 구현을 해야할까?
  - 엔티티에 대한 검증하는 방식은 고민이 많은 부분인 것 같습니다. 저라면 일단 초기화하는 생성자에 위치하거나 수정 하는 메서드에 검증 로직을 위치하고 엔티티 내에 메서드를 추가하는 것보다 내부에 Validator 클래스를 사용하는 것이 좋아보이더라구요.
  - @Valid는 dto에서만 사용하여 최대한 라이브러리 사용을 줄일 것 같네요.!

SeokRae commented on src/main/java/org/programmers/springboardjpa/domain/post/dto/PostRequest.java:
```diff
@@ -0,0 +1,25 @@
+            @NotEmpty(message = "내용은 Empty 일 수 없습니다.")
+            String content,
+            @NotNull
+            UserRequest userDto
```
userDto라는 의미가 author 인걸까요?

ICCHOI replied:
유저에 대한 정보를 가지고 있는 객체입니다!

SeokRae replied:
Post와 관련된 user는 어떤 유저일까요?
이 dto를 파라미터로 제공되는 API를 Front 개발자에게 제공할 때 어떠한 명칭으로 나가게 될까요?
구현하신 프로젝트의 API 명세서를 작성해보시는 것이 좋을 것 같습니다.

ICCHOI replied:
<img width="1005" alt="스크린샷 2022-06-03 오후 3 44 33" src="https://user-images.githubusercontent.com/48702370/171801570-786ad9fb-a2a3-409c-90a3-518667a9f6db.png">
이런 식으로 구성하였는데 userDto의 의미를 좀 더 명시적으로 표현하면 좋겠다는 생각이 드네요!

ynoolee replied:
게시글을 작성하는 과정에서 너무 많은 유저에 대한 정보가 필요한것 같다는 생각이 듭니다. 이렇게 하신 이유가 있으실까용?
글을 작성하는 모든 로그인 하지 않은 사용자의 정보를 createPost를 할 때 마다 매 번 저장하도록 하고 있는 걸까요?

ICCHOI replied:
post에 대한 api만 존재하기 때문에 이런 형태가 된 것 같습니다!

SeokRae commented on src/test/java/org/programmers/springboardjpa/post/service/PostDefaultServiceStubTest.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    @Test
+    @DisplayName("")
```
@DisplayName은 내용이 비어있으면 안됩니다.!

ICCHOI replied:
테스트 케이스 작성 중 아래와 같은 문제가 있어 작성을 보류 했었는데 해결하였습니다 :)

SeokRae commented on src/test/java/org/programmers/springboardjpa/post/service/PostDefaultServiceStubTest.java:
```diff
@@ -0,0 +1,100 @@
+        postResponseList.add(post);
+
+        //when
+        var postList = postService.getPostList(
+                Pageable.ofSize(1).withPage(1));
```
```suggestion
        List<PostResponse.PostResponseDto> postList = postService.getPostList(PageRequest.of(0, 1));
```

저는 보통 PageRequest.of(0, 1) PageRequest 클래스에서 제공하는 팩토리 메서드를 사용하는데 Pageable.ofSize(1).withPage(1)의 의미는 뭘까요?

ICCHOI replied:
Pageable에 대한 구현체를 찾아보다 of 메소드를 통해 PageableImpl이라는 구현체를 통해 반환이 가능하여 테스트를 진행하려고 했었는데 PageRequest를 사용하는 방법이 맞는 것 같네요.
이 부분 추가적으로 학습이 필요할 것 같습니다.

ICCHOI committed:
Config:필요없는 의존성을 제거하라

ICCHOI committed:
test:Test를 수정하라

PostDefaultServiceStubTest의 getAllPost()를 PageRequest.of를 사용하여 작성하라
PostDefaultServiceTest의 필요없는 어노테이션을 삭제하라

ICCHOI committed:
fix:잘못 매핑된 어노테이션을 수정하라

ICCHOI committed:
docs:README.md를 작성하라

ynoolee commented on src/main/java/org/programmers/springboardjpa/domain/model/BaseEntity.java:
```diff
@@ -0,0 +1,23 @@
+    private String createdBy;
+
+    @CreatedDate
+    @Column(name = "created_at", columnDefinition = "TIMESTAMP")
```
이 속성에만 columnDefinition 을 명시하는 이유가 있을까요?

ICCHOI replied:
날짜를 표시할 수 있는 여러 타입중 특정한 타입을 명시해주려는 목적이였습니다!

SeokRae replied:
DateFormatter 로 여러 타입으로 변환 해보시죠. !

ynoolee commented on src/main/java/org/programmers/springboardjpa/domain/post/api/PostApi.java:
```diff
@@ -0,0 +1,44 @@
+@Validated
+@RequestMapping("/api/v1")
+@RestController
+public class PostApi {
```
오 Controller 라는 네이밍에서 벗어난 이름이네요 그렇다면 현재처럼 api 를 위한 것이 아닌 View 를 리턴하는 클래스의 이름은 어떤식으로 작성하실지 궁금합니다!

ICCHOI replied:
ModelAndView를 return 한다면 그대로 Controller라는 이름을 쓸 것 같네요!

ynoolee commented on src/main/java/org/programmers/springboardjpa/domain/post/api/PostApi.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @PostMapping("/posts")
+    public PostResponse.PostResponseDto save(@Valid @RequestBody PostRequest.PostCreateRequestDto createRequest) {
```
클래스레벨에 @Validated 를 사용하고 인자에는 @Valid 를 사용하셨는데 그렇다면 클래스단에 @Validated 를 사용하신 이유가 무엇일까요? 🤔

ICCHOI replied:
인자에서 사용되는 `@Valid`는 해당 어노테이션이 붙은 클래스 안의 검증을 수행하는 어노테이션이며,
`@Validated`는 해당 클래스 내의 Validation을 활성화 하기 위한 어노테이션입니다.
어노테이션을 통하여 id의 `@Positive`를 검증하고자 하였습니다.

SeokRae replied:
Id 값을 직접 사용하는 사이트의 사용 사례(tistory)를 보면 id의 PathVariable의 값이 어떠한 값이던지 page(resource)가 존재하면
그 데이터를 반환하고 그게 아니면 존재하지 않는 페이지에 대한 예외를 던지는 것 같습니다.

결국 제가 이야기하고 싶은 내용은 자원이 존재하는 id 값에 대한 검증을 다 하는게 아니면
**positive로 양수 값만 들어오게끔 처리해야 하는 건지**에 대한 고민이 드는 것 같습니다. 어떻게 생각하시나요?

ynoolee commented on src/main/java/org/programmers/springboardjpa/domain/post/api/PostApi.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @GetMapping("/posts")
+    public List<PostResponse.PostResponseDto> getAllPost(Pageable pageable) {
```
클래스 이름이 PostApi 이기 때문에 해당 클래스 내부의 메소드들은 Post 라는 것을 이름의 일부로 가질 필요가 없을 거라고 생각됩니다! 이미 클래스 이름에서 잘 표현해주시고 있으시다는 생각이 들어요! 그리고 update,save 의 경우는 updatePost ,savePost라고 하지 않으셨듯이 다른 것들도 getAll, get 이라고 해주는 게 어떨까요?

ICCHOI replied:
좋은 의견이네요! 메소드 네이밍에 대한 어떤 기준을 항상 세워놓고 작성하는 습관을 들여야할 것 같아요.

ynoolee commented on src/main/java/org/programmers/springboardjpa/domain/post/domain/Post.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    public boolean equals(Object o) {
+        if (this==o) return true;
```
연산자 앞 뒤로 띄어쓰기가 있으면 가독성이 더 좋아질 것 같습니다!

ICCHOI replied:
자동 생성된 메소드라 그대로 뒀습니다 ㅎㅎ;;

ynoolee commented on src/main/java/org/programmers/springboardjpa/domain/post/domain/Post.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
영속성 컨텍스트에서는 이미 기본적으로 id 값을 사용한 엔티티 객체간의 equality 를 판단하는 것 으로 알고있는데  이렇게 직접 id 를 통한 equals 를 재정의하신 이유가 있으신지 궁금합니다~
저는 개인적으로 equals 의 경우 equals 규약(이펙티브자바 Item10~ Item 11 ) 을 지킨 오버라이딩이 어렵다고 생각되어 인텔리제이나 롬복에서 제공하는 것을 사용하는 편입니다!

ICCHOI replied:
영속성 컨텍스트와 상관 없이 도메인은 항상 equals와 hashcode를 구현해야한다고 생각합니다.
JPA로 작성한 코드는 영속성 컨텍스트에 있는 값을 찾을 때는 1차 캐시를 이용해 같은 id로 같은 값을 찾아올 수 있지만,
1차 캐시를 초기화 한 후 다시 데이터베이스에서 동일한 엔티티를 읽어오는 경우 초기화 이전에 얻었던 객체와 다른 객체가 생성 될 수 있습니다. 또 제가 사용한 Apache Commons의 EqualsBuilder는 Effective Java의 rule을 이미 따르고 있네요!
https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/builder/EqualsBuilder.html

ynoolee commented on src/main/java/org/programmers/springboardjpa/domain/post/dto/PostRequest.java:
```diff
@@ -0,0 +1,25 @@
+import javax.validation.constraints.NotNull;
+
+public class PostRequest {
+    public record PostCreateRequestDto(
```
저도 요즘 고민한 부분인데 DTO 라는 이름을 붙여줘야만 할까? 에 대한 의문을 갖고 있습니다. 개인적으로는 CreatePostRequest 라는 이름만으로 DTO 임을 알 수 있을 것 같다는 생각을 했습니다! 😊

ICCHOI replied:
개인적으로는 네이밍으로 알 수 있는 부분이 많다고 생각해서 DTO로만 활용된다면 이름을 꼭 붙여줘야한다고 생각합니다.
해당 record가 DTO 외의 더 큰 역할을 가지고 있다면 네이밍을 다르게 바꿀 수도 있을 것 같네요!

ynoolee commented on src/main/java/org/programmers/springboardjpa/domain/post/service/PostConverter.java:
```diff
@@ -0,0 +1,46 @@
+                .build();
+    }
+
+    private User convertUser(UserRequest userDto) {
```
PostConverter 인데 User 객체와 관련된 변환로직이 있는 이유가 뭘까용 ?

ynoolee commented on src/main/java/org/programmers/springboardjpa/domain/post/service/PostDefaultService.java:
```diff
@@ -0,0 +1,60 @@
+
+    @Override
+    public PostResponseDto getPost(Long id) {
+        return postRepository.findById(id)
```
인창님은 스트림의 map 을 잘 사용하시는군요! 저도 map 을 활용하는 코드를 작성해봐야겠습니다 ㅠㅠ저는 orElseThrow 밖에 안하고 있었네요..

ynoolee commented on src/main/java/org/programmers/springboardjpa/domain/user/dto/UserDto.java:
```diff
@@ -0,0 +1,26 @@
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Positive;
+
+public class UserDto {
```
Post 에서는 요청과 응답에 대한 DTO 를 분리해주시고 여기서는 한 곳에 둔 것은, 요청과 응답에 대한 DTO 의 수가 분리할 만큼 많지 않기 때문일까요 🥺 ?

ICCHOI replied:
네 맞습니다!!

ynoolee commented on src/main/java/org/programmers/springboardjpa/global/error/ErrorResponse.java:
```diff
@@ -0,0 +1,85 @@
+        this.code = code.getCode();
+    }
+
+    private ErrorResponse(final ErrorCode code) {
```
생성자 오버로딩을 고려해보시면 어떨까요?

ICCHOI replied:
불필요한 객체 생성을 제한하기 위해서 of 메소드를 통해서만 생성하도록 구현하였습니다!

ynoolee reviewed (commented):
인창님 수고 많으셨습니다~😊

JiyoungMa commented on src/main/java/org/programmers/springboardjpa/domain/post/service/PostConverter.java:
```diff
@@ -0,0 +1,46 @@
+        );
+    }
+
+    private UserResponse convertUserDto(User user) {
+        return new UserResponse(
+                user.getId(),
+                user.getName(),
+                user.getAge(),
+                user.getHobby()
+        );
```
이 부분 또한 연우님이 말씀하신 것 처럼 User와 관련된 부분인데 왜 Post Converter에 있는지 궁금합니다!

ICCHOI replied:
POST API만 존재하기 때문에 UserConverter를 따로 만들지 않았습니다!

JiyoungMa commented on src/test/java/org/programmers/springboardjpa/post/controller/PostApiTest.java:
```diff
@@ -0,0 +1,254 @@
+                        "인터넷 서핑"
+                )
+        );
+        ResultActions result = this.mockMvc.perform(
+                post("/api/v1/posts")
+                        .content(this.objectMapper.writeValueAsString(postCreateRequestDto))
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .accept(MediaType.APPLICATION_JSON)
+        );
+
+        //then
+        result.andExpect(status().isOk())
+                .andDo(print())
+                .andDo(document("posts-save",
+                        requestFields(
+                                fieldWithPath("title").type(JsonFieldType.STRING).description("제목"),
+                                fieldWithPath("content").type(JsonFieldType.STRING).description("내용"),
+                                fieldWithPath("userDto").type(JsonFieldType.OBJECT).description("유저"),
+                                fieldWithPath("userDto.name").type(JsonFieldType.STRING).description("유저 이름"),
+                                fieldWithPath("userDto.age").type(JsonFieldType.NUMBER).description("유저 나이"),
+                                fieldWithPath("userDto.hobby").type(JsonFieldType.STRING).description("유저 취미")
+                        )
+                        ,
+                        responseFields(
+                                fieldWithPath("id").type(JsonFieldType.NUMBER).description("아이디"),
+                                fieldWithPath("title").type(JsonFieldType.STRING).description("제목"),
+                                fieldWithPath("content").type(JsonFieldType.STRING).description("내용"),
+                                fieldWithPath("userDto").type(JsonFieldType.OBJECT).description("유저"),
+                                fieldWithPath("userDto.id").type(JsonFieldType.NUMBER).description("유저 아이디"),
+                                fieldWithPath("userDto.name").type(JsonFieldType.STRING).description("유저 이름"),
+                                fieldWithPath("userDto.age").type(JsonFieldType.NUMBER).description("유저 나이"),
+                                fieldWithPath("userDto.hobby").type(JsonFieldType.STRING).description("유저 취미")
+                        )
+                ));
```
두 부분으로 나누어서 진행하신 이유는 when과 then 부분을 구분하기 위해서이실까요?

ICCHOI replied:
넵 정확합니다!!!

JiyoungMa commented on src/test/java/org/programmers/springboardjpa/post/controller/PostApiTest.java:
```diff
@@ -0,0 +1,254 @@
+        );
+
+        //then
+        result.andExpect(status().isOk())
```
상태 코드를 제외한 부분은 따로 확인을 안 하신 이유가 있으신지 궁금합니다! 이미 postResponseDto도 있어서 asssert까지 진행하면 더 좋지 않을까?라는 개인적인 생각입니다😊

ICCHOI replied:
그렇네요...! 좋은 의견 감사합니다

JiyoungMa commented on src/test/java/org/programmers/springboardjpa/post/controller/PostApiTest.java:
```diff
@@ -0,0 +1,254 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@AutoConfigureRestDocs
+@WebMvcTest(PostApi.class)
```
@SpringBootTest 말고 @WebMvcTest를 사용하신 이유가 궁금합니다!

ICCHOI replied:
mockMvc를 통해서 controller layer만 분리하여 테스트를 진행하기 위해서  WebMvcTest를 사용하였습니다 :)

JiyoungMa commented on src/test/java/org/programmers/springboardjpa/post/controller/PostApiTest.java:
```diff
@@ -0,0 +1,254 @@
+        result.andExpect(status().isOk())
+                .andDo(print())
+                .andDo(document("posts-get",
+                        pathParameters(
+                                parameterWithName("id").description("게시글 아이디")
+                        ),
```
path parameter도 rest doc에 추가할 수 있군요?👍

JiyoungMa commented on src/test/java/org/programmers/springboardjpa/post/service/PostDefaultServiceStubTest.java:
```diff
@@ -0,0 +1,99 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SpringBootTest
+class PostDefaultServiceStubTest {
```
Service에 대해 Stub과 Mock 두 테스트를 모두 진행하신 이유가 궁금합니다!

ICCHOI replied:
행위 검증과 상태 검증 두 케이스를 모두 테스트하고 싶었습니다 :)

JiyoungMa commented on src/test/java/org/programmers/springboardjpa/post/service/PostDefaultServiceStubTest.java:
```diff
@@ -0,0 +1,99 @@
+        var updatedPost = postService.getPost(savedPost.id());
+
+        //then
+        assertThat(updatedPost.id()).isEqualTo(savedPost.id());
+        assertThat(updatedPost.title()).isEqualTo(requestDto.title());
+        assertThat(updatedPost.content()).isEqualTo(requestDto.content());
```
새로운 객체를 만들지 않고 하나하나 assertThat을 하신 이유가 혹시 따로 있으신가요?

ICCHOI replied:
dto에는 id가 존재하지 않아서 id는 따로 비교 해야해서 각각 비교를 했었습니다!

JiyoungMa reviewed (commented):
인창님 고생하셨습니다! 고민하신 부분이 많이 보여서 저도 많이 참고해보려고 합니다!👍

ICCHOI committed:
feat:ConstraintViolationException handle method를 구현하라

ICCHOI committed:
refactor:Valid 어노테이션을 Validated로 변경하라

ICCHOI committed:
refactor:var을 Post로 변경하라

ICCHOI committed:
refactor:생성자를 builder로 변경하라

ICCHOI committed:
refactor:생성자 오버로딩을 적용하라

ICCHOI committed:
test:printDo()를 제거하라

ICCHOI committed:
test:테스트의 네이밍을 수정하라

ICCHOI committed:
refactor:service를 스트림으로 변경하라

ICCHOI committed:
refactor:naming과 간격을 수정하라

ICCHOI committed:
Create codeball.yml

ICCHOI committed:
Delete .github/workflows directory
